### PR TITLE
[DO NOT MERGE] Display meaningful module traits

### PR DESCRIPTION
**Pull request by wvu**
# DO NOT MERGE

```
msf5 exploit(linux/upnp/belkin_wemo_upnp_exec) > info

       Name: Belkin Wemo UPnP Remote Code Execution
     Module: exploit/linux/upnp/belkin_wemo_upnp_exec
   Platform: Unix, Linux
       Arch: cmd, mipsle
 Privileged: Yes
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2014-04-04

Provided by:
  phikshun
  wvu <wvu@metasploit.com>
  nstarke

Available targets:
  Id  Name
  --  ----
  0   Unix In-Memory
  1   Linux Dropper

Check supported:
  Yes

Basic options:
  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  Proxies                   no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS                    yes       The target address range or CIDR identifier
  RPORT    49152            yes       The target port (TCP)
  SRVHOST  0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
  SRVPORT  8080             yes       The local port to listen on.
  SSL      false            no        Negotiate SSL/TLS for outgoing connections
  SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
  URIPATH                   no        The URI to use for this exploit (default is random)
  VHOST                     no        HTTP server virtual host

Payload information:

Description:
  This module exploits a command injection in the Belkin Wemo UPnP API
  via the SmartDevURL argument to the SetSmartDevInfo action. This
  module has been tested on a Wemo-enabled Crock-Pot, but other Wemo
  devices are known to be affected, albeit on a different RPORT
  (49153).

References:
  https://web.archive.org/web/20150901094849/http://disconnected.io/2014/04/04/universal-plug-and-fuzz/
  https://github.com/phikshun/ufuzz
  https://gist.github.com/phikshun/10900566
  https://gist.github.com/phikshun/9984624
  https://www.crock-pot.com/wemo-landing-page.html
  https://www.belkin.com/us/support-article?articleNum=101177
  http://www.wemo.com/

Module stability:
  Module should not crash the service

Module reliability:
  Module is expected to get a shell every time it fires

Module side effects:
  Module leaves payload or a dropper on the target machine

CVE not available for the following reason:
  Patched in 2.00.8643 without vendor disclosure

msf5 exploit(linux/upnp/belkin_wemo_upnp_exec) >
```

Note that I did reorder module traits to stability, reliability, and then side effects due to the following thought process:

1. Will I crash anything if I run this?
2. Can I expect to get a shell?
3. What might need to be cleaned up?

#10707, #11769, #11820

### CODE REVIEW

**Comment by wvu-r7** [comment]
This is ready for review.

File: lib/msf/base/serializer/readable_text.rb [hunk 1]
```diff
@@ -514,19 +516,21 @@ def self.dump_basic_module(mod, indent = '')
...
-
     # Description
     output << "Description:\n"
     output << word_wrap(Rex::Text.compress(mod.description))
     output << "\n"
 
+    # References
     output << dump_references(mod, indent)
 
-    output << "\n"
```
Comment (wvu-r7) [inline lib/msf/base/serializer/readable_text.rb]: Curiously, this was actually an extra blank line.

Comment (wvu-r7) [reply lib/msf/base/serializer/readable_text.rb]:
```
[1] pry(#<Msf::Modules::Exploit__Linux__Upnp__Belkin_wemo_upnp_exec::MetasploitModule>)> puts Msf::Serializer::ReadableText.dump_basic_module(self, '  ')

       Name: Belkin Wemo UPnP Remote Code Execution
     Module: exploit/linux/upnp/belkin_wemo_upnp_exec
   Platform: Unix, Linux
       Arch: cmd, mipsle
       Rank: Excellent

Provided by:
  phikshun
  wvu <wvu@metasploit.com>
  nstarke

Module side effects:
 artifacts-on-disk

Module stability:
 crash-safe

Module reliability:
 repeatable-session

Description:
  This module exploits a command injection in the Belkin Wemo UPnP API
  via the SmartDevURL argument to the SetSmartDevInfo action. This
  module has been tested on a Wemo-enabled Crock-Pot, but other Wemo
  devices are known to be affected, albeit on a different RPORT
  (49153).

References:
  CVE: Not available
  https://web.archive.org/web/20150901094849/http://disconnected.io/2014/04/04/universal-plug-and-fuzz/
  https://github.com/phikshun/ufuzz
  https://gist.github.com/phikshun/10900566
  https://gist.github.com/phikshun/9984624
  https://www.crock-pot.com/wemo-landing-page.html
  https://www.belkin.com/us/support-article?articleNum=101177
  http://www.wemo.com/


=> nil
[2] pry(#<Msf::Modules::Exploit__Linux__Upnp__Belkin_wemo_upnp_exec::MetasploitModule>)>
```

**Comment by busterb** [comment]
It looks like this probably also changes how the metadata is stored in the metadata file. Is that alright? E.g. this stuff would change:

https://github.com/rapid7/metasploit-framework/commit/8a218f60a9dde80b0b50395e6f6bdb3814397126#diff-d77d5e503ad1439f585ac494268b351b

**Comment by wvu-r7** [comment]
I think it can be a problem in the long run if we desire programmatic selection of module traits. We can't rely on the description strings to remain static. I have an idea I'm testing.

**Comment by sgonzalez-r7** [comment]
I'm starting to consume [modules_metadata_base.json](https://github.com/rapid7/metasploit-framework/blob/master/db/modules_metadata_base.json) programmatically, i.e. parse the json into a data structure for processing.  Having a reliable, fixed structure is important for machine consumption.  Take the traits properties inside `notes`
```
"notes": {
  "AKA": [
    "SA-CORE-2019-003"
  ],
  "Stability": [
    "crash-safe"
  ],
  "SideEffects": [
    "ioc-in-logs"
  ],
  "Reliability": [
    "unreliable-session"
  ]
}
```
Regardless of where the traits live, e.g. inside `notes` or promoted to the top-level, having them be of the same structure would be very helpful, e.g. each trait is an array of strings.
```
"name": [
  "description-1",
  "description-2",
  ...
]
```
What are thoughts on moving towards a standardization of the json structure?  Maybe defining a [json schema](https://json-schema.org/) to spec it out.

**Comment by wvu-r7** [comment]
I've separated out the `NOCVE` changes into #11820.

**Comment by wvu-r7** [comment]
Rebased for clarity after #11820 merged.

**Comment by wvu-r7** [comment]
Rebased to pull in changes from #11864.

**Comment by bcoles** [comment]
Were module traits ever documented anywhere outside of the source code and PRs (#10707 #10744) ?

**Comment by wvu-r7** [comment]
@bcoles: Not to my knowledge.

**Comment by wvu-r7** [comment]
Closing this due to conflicts and requiring a serious refactor. I'll open a new PR when the time comes.

**Comment by wvu-r7** [comment]
Need to revisit this. Marking as `attic`.

**Comment by adfoster-r7** [comment]
@wvu-r7 I like this! Definitely makes this information more visible IMO

> Closing this due to conflicts and requiring a serious refactor. I'll open a new PR when the time comes.

What conflicts/refactoring is this referring to specifically? 👀

**Comment by wvu-r7** [comment]
I think this: https://github.com/rapid7/metasploit-framework/pull/11785#issuecomment-488790403. That's why I didn't map the constants directly to descriptive strings. I kept the self-descriptive strings for the module cache?

**Pull request closed without merging**